Modify deploy_docs script to include Android

danielnugraha opened the pull request:
<!--
Thank you for opening a pull request (PR)!

Contribution guidelines: https://github.com/adap/flower/blob/main/CONTRIBUTING.md
-->

## Issue

### Description

<!--
Describe the problem addressed by this PR.

Example: The variable name `rnd` could be misinterpreted as an abbreviation of *random*, but it refers to the current server round.
-->

The Android SDK doesn't have any API reference, so it's great to have one to soften the learning curve for users using our API.

### Related issues/PRs

<!--
Link issues and/or PRs that are related to this PR.

Example: Fixes #123. See also #456 and #789.
-->

## Proposal

### Explanation

<!--
Explain the changes and how they improve the issue described above.

Example: The variable `rnd` was renamed to `server_round` to improve readability.
-->

Build setup to generate API reference in html, basic styling to conform to Flower design guidelines and logo change are applied. Also the deploy_docs.sh is updated to generate android API reference. Changes base on Android SDK, merge only after Android SDK is merged.

<img width="1512" alt="Screenshot 2023-08-28 at 08 47 44" src="https://github.com/adap/flower/assets/52073299/72b08e1f-f449-4573-961b-4f712f5506a3">


<img width="1511" alt="Screenshot 2023-08-28 at 08 48 24" src="https://github.com/adap/flower/assets/52073299/c7aeab44-b0f5-49e7-b86f-e5dab2573192">

### Checklist

- [ ] Implement proposed change
- [ ] Write tests
- [ ] Update [documentation](https://flower.dev/docs/writing-documentation.html)
- [ ] Update [changelog](https://github.com/adap/flower/blob/main/doc/source/changelog.rst)
- [ ] Make CI checks pass
- [ ] Ping maintainers on [Slack](https://flower.dev/join-slack/) (channel `#contributions`)

### Any other comments?

<!--
Please be aware that it may take some time until the maintainers can review the PR.
Smaller PRs with good descriptions can be considered much more easily.

If you have an urgent request or question, please use the Flower Slack:

    https://flower.dev/join-slack/ (channel: #contributions)

Thank you for contributing to Flower!
-->

tanertopal commented on src/kotlin/flwr/.gitignore:
```diff
@@ -0,0 +1 @@
+/build
```
Move this into the global gitignore

danielnugraha replied:
This has been moved in add-android-sdk, I'll pull the branch here.

tanertopal commented on dev/deploy-docs.sh:
```diff
@@ -40,3 +40,10 @@ cd examples/doc
 make docs
 cd build/html
 aws s3 sync --delete --exclude ".*" --exclude "v/*" --acl public-read --cache-control "no-cache" ./ s3://flower.dev/docs/examples
+
+# Build and deploy Flower Android docs
+cd $ROOT
+cd src/kotlin
+./gradlew dokkaHtml
+cd build/dokka/html
+aws s3 sync --delete --exclude ".*" --exclude "v/*" --acl public-read --cache-control "no-cache" ./ s3://flower.dev/docs/android
```
Similar to Swift lets separate the script so it can be used in a dedicated CI job

tanertopal commented on src/kotlin/.gitignore:
```diff
@@ -0,0 +1,15 @@
+*.iml
+.gradle
+/local.properties
+/.idea/caches
+/.idea/libraries
+/.idea/modules.xml
+/.idea/workspace.xml
+/.idea/navEditor.xml
+/.idea/assetWizardSettings.xml
+.DS_Store
+/build
+/captures
+.externalNativeBuild
+.cxx
+local.properties
```
Let's move this to the global level.

tanertopal commented on .github/workflows/android-release.yml:
```diff
@@ -0,0 +1,24 @@
+          AWS_ACCESS_KEY_ID: ${{ secrets.AWS_ACCESS_KEY_ID }}
+          AWS_SECRET_ACCESS_KEY: ${{ secrets. AWS_SECRET_ACCESS_KEY }}
+        run: ./dev/deploy-android-docs.sh
+          
```
```suggestion

```

tanertopal commented on .github/workflows/android-release.yml:
```diff
@@ -0,0 +1,23 @@
+          AWS_DEFAULT_REGION: ${{ secrets. AWS_DEFAULT_REGION }}
+          AWS_ACCESS_KEY_ID: ${{ secrets.AWS_ACCESS_KEY_ID }}
+          AWS_SECRET_ACCESS_KEY: ${{ secrets. AWS_SECRET_ACCESS_KEY }}
+        run: ./dev/deploy-android-docs.sh
```
```suggestion
        run: ./dev/deploy-android-docs.sh

```

tanertopal commented on dev/deploy-android-docs.sh:
```diff
@@ -0,0 +1,28 @@
+cd src/kotlin
+./gradlew dokkaHtml
+cd flwr/build/dokka/html
+aws s3 sync --delete --exclude ".*" --exclude "v/*" --acl public-read --cache-control "no-cache" ./ s3://flower.dev/docs/android
```
```suggestion
aws s3 sync --delete --exclude ".*" --acl public-read --cache-control "no-cache" ./ s3://flower.dev/docs/android
```

danielnugraha commented on .github/workflows/android-release.yml:
```diff
@@ -0,0 +1,23 @@
+name: Android Client Publish
+
+on:
+  pull_request
```
Do we always wants to deploy the docs on pull request?

tanertopal commented:
@danielnugraha, regarding the renaming of the CI YAML file, I just want to ensure that each team knows which YAML file belongs to their work. It could be clearer right now. Having just `android.yml` is, for me, as we have a mono repo, and the number of YAML files is getting out of control.

wrt to your comment about when to deploy the Android docs, you mentioned the following two options (correct me if I misread):

1. Deploy on each commit to `main`
2. Deploy on each release

I am undecided. In the Python code base, we used to deploy on each push to merge and switched now to having versioned docs + latest docs.

I am okay with deploying only on each release for Android. In the case of the iOS docs which we just merged we deploy on each commit to `main` - although the swift release is in sync with that right?

danielnugraha commented:
> @danielnugraha, regarding the renaming of the CI YAML file, I just want to ensure that each team knows which YAML file belongs to their work. It could be clearer right now. Having just `android.yml` is, for me, as we have a mono repo, and the number of YAML files is getting out of control.
> 
> wrt to your comment about when to deploy the Android docs, you mentioned the following two options (correct me if I misread):
> 
>     1. Deploy on each commit to `main`
> 
>     2. Deploy on each release
> 
> 
> I am undecided. In the Python code base, we used to deploy on each push to merge and switched now to having versioned docs + latest docs.
> 
> I am okay with deploying only on each release for Android. In the case of the iOS docs which we just merged, we deploy on each commit to `main` - although the swift release is in sync with that, right?
> 
> @danieljanes any thoughts on this?

I've changed the script name to android.yml, should I separate the publishing and deploying the docs? The publishing can be complicated if it triggers on every pull request merge.

tanertopal commented on .github/workflows/android.yml:
```diff
@@ -0,0 +1,52 @@
+          AWS_DEFAULT_REGION: ${{ secrets. AWS_DEFAULT_REGION }}
+          AWS_ACCESS_KEY_ID: ${{ secrets.AWS_ACCESS_KEY_ID }}
+          AWS_SECRET_ACCESS_KEY: ${{ secrets. AWS_SECRET_ACCESS_KEY }}
+        run: ./dev/deploy-android-docs.sh
+          
```
```suggestion
        run: ./dev/deploy-android-docs.sh

```
